### ISSUE TITLE
Feature request: make Connection implement Send + Sync

### ISSUE THREAD

**Issue by dbkaplun** [jgallagher/rusqlite#342]

**Comment by gwenn** [comment]
You should read #188.

**Comment by benesch** [comment]
Apologies, but I don't understand what in the discussion in #188 precludes this. Based on my read of the SQLite documentation, if SQLite is opened with the flag SQLITE_OPEN_FULLMUTEX (serialized mode), then it should be perfectly safe to share the sqlite object between threads, i.e., by implementing `Sync`. I get that the Rust wrapper at the moment might need to be reworked, but I don't think anyone has yet pointed out a technical reason why it *couldn't* be made to work?

**Comment by dimitarvp** [comment]
I just recently stumbled upon this. Given that sqlite3's default mode is serialized then I see no reason why shouldn't I be able to have `RwLock<rusqlite::Connection>` type which I will just give away to any thread that wants a connection to that particular DB, and let sqlite3 do the mutex locking/unlocking operations.

But, given that sqlite3 indeed also does mutex lock/unlock then I guess it's fine if we just use a connection pool like `r2d2_sqlite` or, in my case -- since I am writing an sqlite3 Elixir wrapper for `rusqlite` -- just use a registry that maps threads to DB connections.

So I'd also love a `RwLock`-enabled `rusqlite::Connection` but the alternatives seem okay-ish.

**Comment by gwenn** [comment]
`rusqlite` is designed such as `Connection` are not `Sync`.
Based on Rust guarantees, the `sqlite3` handle can be set to `multi-thread` mode (instead of `serialized` mode).
It seems wise because some API are not threadsafe even in `serialized` mode:
- https://sqlite.org/c3ref/errcode.html
- https://sqlite.org/c3ref/last_insert_rowid.html

**Comment by benesch** [comment]
> `rusqlite` is designed such as `Connection` are not `Sync`.
> Based on Rust guarantees, the `sqlite3` handle can be set to `multi-thread` mode (instead of `serialized` mode).
> It seems wise because some API are not threadsafe even in `serialized` mode:
> 
> * https://sqlite.org/c3ref/errcode.html
> * https://sqlite.org/c3ref/last_insert_rowid.html
> * https://sqlite.org/c3ref/changes.html

Ooh, ok, that's very compelling. Thanks for that. I retract my previous comment! Seems like a `Mutex<rusqlite::Connection>` is the way to go.

**Comment by dimitarvp** [comment]
> Thanks for that. I retract my previous comment! Seems like a Mutex<rusqlite::Connection> is the way to go.

Yeah, that's what I ended up doing, now just pondering what kind of pooling should I use. Rest seems clear enough now.

**Comment by gwenn** [comment]
Here are other API not threadsafe:
https://sqlite.org/c3ref/column_database_name.html
> If two or more threads call one or more [column metadata interfaces](https://sqlite.org/c3ref/column_database_name.html) for the same [prepared statement](https://sqlite.org/c3ref/stmt.html) and result column at the same time then the results are undefined.

(found while checking #1672)
